File: trunk/swisscenter/ext/broadcast/broadcast.pl
#!/usr/bin/perl

# Script to ensure that the Pinnacle Showcenter (and possibly all other Syabas iHome
# devices..) finds this swisscenter server. It works by detecting the upnp::rootdevice
# NOTIFY announcement for a SYABAS device and sending a URL request. I can't find any
# documentation on this so it's just been unpicked using ethereal.

# If you receive an error that "IO::Socket::Multicast" doesn't exist, then install it
# as follow (as the root user):
#   perl -MCPAN -e 'install IO:Interface'
#   perl -MCPAN -e 'install IO::Socket::Multicast'
#   perl -MCPAN -e 'install HTTP::Lite'

# Bring in a couple of modules
use IO::Socket::Multicast;
use HTTP::Lite;
use Getopt::Long;

# extract hostname and IP from the command line
my $host = "SwissCenter";
my $addr = "192.168.0.3:8080";
my $debug;

$result = GetOptions("host=s" => \$host, "ip=s" => \$addr, "debug" => \$debug);
print "H=$host I=$addr \n" if ($debug);

# the upnp multicast group:port
use constant GROUP => '239.255.255.250';
use constant PORT  => '1900';

# and create the multicast endpoint
my $sock = IO::Socket::Multicast->new(Proto=>'udp',LocalPort=>PORT,ReuseAddr=>1);
$sock->mcast_add(GROUP) || die "Couldn't set multicasp UPnP Group: $!\n";

# should really listen for a command to stop
while (1) {
    my $data;
    # read a multicast message
    $hispaddr = $sock->recv($data,1024);
    ($port, $hisiaddr) = sockaddr_in($hispaddr);
    $peer_addr = inet_ntoa($hisiaddr);
 
    #print "Recv from ".$peer_addr.":".$hisaddr.":".$port.":".$data;

    if ($data =~ /NOTIFY/) {
        # Found a SYSBAS myiHome Device
        if ($data =~ /USN: uuid:myiBoxUPnP::upnp:rootdevice/) {
            # Find out where it is
            $data =~ /LOCATION: (.*)/;
            $url=$1;
            sleep 1;
            $http = new HTTP::Lite;
            $req = $http->request($url."?POSTSyabasiBoxURLpeername=$host&peeraddr=$addr&")
              or die "Unable to get document: $!";
            print $http->body() if ($debug);
        }
    }
}

File: ext/broadcast/broadcast.py
#!/usr/bin/python

import socket
import struct
import re
import urllib
from optparse import OptionParser

parser = OptionParser()
parser.add_option("--host", dest = "host", default = "SwissCenter", help = "Set the host identifier")
parser.add_option("--ip", dest="addr", default = "192.168.0.3:8080", help = "Set the ip address")
parser.add_option("--debug", dest="debug", action="store_true", default = False, help = "Enable debugging information")
parser.add_option("--match", dest="match", default = "USN: uuid:myiBoxUPnP::upnp:rootdevice", help = "Set the packet match string")
(options, args) = parser.parse_args()

# Setup debug logging
def disabled(func):
  def empty(*args, **kargs):
    pass
  return empty
def enabled(func):
  return func

state = (disabled,enabled)[int(options.debug)]
@state
def debugPrint(s):
  print s

# UPnP group / port
group = "239.255.255.250";
port = 1900;

debugPrint("H=%s I=%s M=\"%s\"" % (options.host,options.addr,options.match))

# Create multicast socket
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
sock.bind(('',port))
mcast = struct.pack("4sl", socket.inet_aton(group), socket.INADDR_ANY)
sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mcast)

while True:

  # Read packet
  (data, source) = sock.recvfrom(1024)

  debugPrint("Data from (%s:%d):\n%s" % (source[0], source[1], data))

  if re.match("^NOTIFY",data):
    # Found a SYABAS myIHome Device
    if re.search(options.match, data):
      # Extract URL
      url = re.search("LOCATION: (.*)\r", data)

      try:
        target = "%s?POSTSyabasiBoxURLpeername=%s&peeraddr=%s&" % (url.group(1), options.host, options.addr)
        request = urllib.urlopen(target).read()
        debugPrint(request)
      except:
        pass
